env: use mydef to template mpicc scripts

hzhou opened the pull request:
## Pull Request Description
We maintain following scripts:
```
src/env/mpicc.sh.in
src/env/mpicc.bash.in
src/env/mpicxx.sh.in
src/env/mpicxx.bash.in
src/env/mpifort.sh.in
src/env/mpifort.bash.in
```

These files contain much redundancy but with subtle differences. They are tedious to maintain and easy to introduce inconsistencies. 

The PR introduce MyDef, a general purpose macro/template system, to allow write structured templates. As result, we use a single `mpicc.def` to replace the old 6 files. Other than whitespace diffs, the template should retain original mechanism and not introduce functional changes.

## `mpicc.sh`, `mpifort.sh`, `mpicxx.sh`
The quoting in POSIX shell versions simply doesn't work.  In
```
allargs="$a $b $c"
$CC $allargs
```
if any of the `$a`, `$b`, `$c` contains spaces or quotes, the shell expansion for the final command will only perform argument breaking at space regardless of quotes or escapes and pass quotes as literal quotes.

Thus we remove the quoting in this PR. We hope users will either have bash or won't deal with arguments that contains spaces or quotes.

## Diffs
To ensure we are not introducing accidental changes, following is the `diff` of the scripts generated by `mpicc.def` and the previous replaced scripts. As shown, they are mostly white space differences.
```diff

diff -b new/src/env/mpicc.bash.in prev/src/env/mpicc.bash.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
203a210,211
>     *)
>     ;;
228a237
> # -----------------------------------------------------------------------
271a281,282
> # -----------------------------------------------------------------------
> #
282d292
<
295d304
<

diff -b new/src/env/mpicxx.bash.in prev/src/env/mpicxx.bash.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
94a99,100
> allargs=("$@")
> argno=0
100,101d105
< allargs=("$@")
< argno=0
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
208a215,216
>     *)
>     ;;
233a242
> # -----------------------------------------------------------------------
276a286,287
> # -----------------------------------------------------------------------
> #
287d297
<
300d309
<

diff -b new/src/env/mpifort.bash.in prev/src/env/mpifort.bash.in
1d0
<
19d17
<
27a26
> # --------------------------------------------------------------------------
55a55
>
56a57
>
76a78
> #---------------------------------------------------------------------
99a102,103
> #
> # ------------------------------------------------------------------------
108a113
>
109a115,116
> allargs=("$@")
> argno=0
113,114d119
< allargs=("$@")
< argno=0
118a124
>       # ----------------------------------------------------------------
123a130
>       # ----------------------------------------------------------------
199d205
<
241a248
>     # Otherwise, just accept the argument
242a250
>     # - end of special handling for .F files
267a276
> # -----------------------------------------------------------------------
270a280
>
292a303
>
343d353
<
360d369
<

diff -b new/src/env/mpicc.sh.in prev/src/env/mpicc.sh.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
97c101
< show_info=""
---
> showinfo=""
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
203a210
>         # -----------------------------------------------------------------
235a243
> # -----------------------------------------------------------------------
278a287,288
> # -----------------------------------------------------------------------
> #
289d298
<
302d310
<
diff -b new/src/env/mpicxx.sh.in prev/src/env/mpicxx.sh.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
94a99
> allargs=""
100d104
< allargs=""
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
208a215
>         # -----------------------------------------------------------------
240a248
> # -----------------------------------------------------------------------
283a292,293
> # -----------------------------------------------------------------------
> #
294d303
<
307d315
<

diff -b new/src/env/mpifort.sh.in prev/src/env/mpifort.sh.in
1d0
<
19d17
<
27a26
> # --------------------------------------------------------------------------
55a55
> #
56a57
>
76a78
> #---------------------------------------------------------------------
99a102,103
> #
> # ------------------------------------------------------------------------
108a113
> #
110c115
< cppflags=()
---
> allargs=""
113d117
< allargs=""
118a123
>       # ----------------------------------------------------------------
123a129
>       # ----------------------------------------------------------------
198a205
>         # -----------------------------------------------------------------
218d224
<
260a267
>     # Otherwise, just accept the argument
261a269,270
>     # - end of special handling for .F files
>
287a297
> # -----------------------------------------------------------------------
308a319
>
359d369
<
376d385
<
diff -b new/src/env/mpicxx.sh.in prev/src/env/mpicxx.sh.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
94a99
> allargs=""
100d104
< allargs=""
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
208a215
>         # -----------------------------------------------------------------
240a248
> # -----------------------------------------------------------------------
283a292,293
> # -----------------------------------------------------------------------
> #
294d303
<
307d315
<

diff -b new/src/env/mpifort.sh.in prev/src/env/mpifort.sh.in
1d0
<
19d17
<
27a26
> # --------------------------------------------------------------------------
55a55
> #
56a57
>
76a78
> #---------------------------------------------------------------------
99a102,103
> #
> # ------------------------------------------------------------------------
108a113
> #
110c115
< cppflags=()
---
> allargs=""
113d117
< allargs=""
118a123
>       # ----------------------------------------------------------------
123a129
>       # ----------------------------------------------------------------
198a205
>         # -----------------------------------------------------------------
218d224
<
260a267
>     # Otherwise, just accept the argument
261a269,270
>     # - end of special handling for .F files
>
287a297
> # -----------------------------------------------------------------------
308a319
>
359d369
<
376d385
<

diff -b new/src/env/mpicxx.sh.in prev/src/env/mpicxx.sh.in
1d0
<
18d16
<
26a25
> # --------------------------------------------------------------------------
41a41
>
61a62
> #---------------------------------------------------------------------
84a86,87
> #
> # ------------------------------------------------------------------------
93a97
>
94a99
> allargs=""
100d104
< allargs=""
105a110
>       # ----------------------------------------------------------------
110a116
>       # ----------------------------------------------------------------
208a215
>         # -----------------------------------------------------------------
240a248
> # -----------------------------------------------------------------------
283a292,293
> # -----------------------------------------------------------------------
> #
294d303
<
307d315
<

diff -b new/src/env/mpifort.sh.in prev/src/env/mpifort.sh.in
1d0
<
19d17
<
27a26
> # --------------------------------------------------------------------------
55a55
> #
56a57
>
76a78
> #---------------------------------------------------------------------
99a102,103
> #
> # ------------------------------------------------------------------------
108a113
> #
110c115
< cppflags=()
---
> allargs=""
113d117
< allargs=""
118a123
>       # ----------------------------------------------------------------
123a129
>       # ----------------------------------------------------------------
198a205
>         # -----------------------------------------------------------------
218d224
<
260a267
>     # Otherwise, just accept the argument
261a269,270
>     # - end of special handling for .F files
>
287a297
> # -----------------------------------------------------------------------
308a319
>
359d369
<
376d385
<
```


[skip warnings]
## Reference
* https://github.com/hzhou/mydef_boot
* https://github.com/hzhou/mydef

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou committed:
autogen: generate mpicc scripts from mpicc.def

Use a mydef template to replace all mpicc template files.

hzhou commented:
test:mpich/warnings/auto

hzhou commented:
test:mpich/warnings/auto

hzhou commented:
test:mpich/submodules

raffenet reviewed (commented):
I've thought about this some and have a few questions.
1. Can the .bash versions of these scripts be removed? Do they offer anything that the .sh version cannot do? I know we've discussed this in the past, but I do not recall the details.
2. Why not just generate `mpicc` and friends from a Python script directly? For instance using triple-quoted strings with substitutions. I am probably missing some complexity that MyDef solves.
3. Is the plan to expand MyDef usage to other areas MPICH if this is accepted?
4. Do downstream maintainers modify these scripts in their distributions? If so, will this change interfere with those modifications? I suppose they can workaround by just modifying the already generated files and adjusting their build scripts.

My overall take is that these scripts are rarely updated, and templating them in some way is a good idea. That said, MyDef is unfamiliar, meaning the templates will likely have a sole maintainer in practice. This is essentially the state of all the code generation scripts in MPICH, so it's something we are currently able to tolerate.

hzhou commented:
> I've thought about this some and have a few questions.
> 
>     1. Can the .bash versions of these scripts be removed? Do they offer anything that the .sh version cannot do? I know we've discussed this in the past, but I do not recall the details.
> 
The only reason we have two versions is because `bash` supports array and `sh` does not. Without array, sh versions have to quote the args and the quoting is hackish and known to be not 100% correct.

Also note, it is not just between sh and bash, there is high degree of redundancy between `mpicc`, `mpicxx`, and `mpifort`, whose variations are less trivial. In fact, I believe we have introduced inconsistency between these scripts. More often we fixed or enhanced something in `mpicc`, then neglected `mpicxx` and `mpifort`.

>     2. Why not just generate `mpicc` and friends from a Python script directly? For instance using triple-quoted strings with substitutions. I am probably missing some complexity that MyDef solves.

If it is just strings with substitutions, then shell variables probably will do. In fact, we have used shell variables as much as we can. There are more subtleties that a python script will introduce more logics than the subtleties themselves and we'll end up maintaining both python scripts and shell scripts. We already do this with bindings scripts. With binding scripts, the saving from duplicate reductions (I think) is still worth it. But here the benefit/harm is not so clear.

> 
>     3. Is the plan to expand MyDef usage to other areas MPICH if this is accepted?

In my personal opinion, MyDef provides superior solutions to most of the scripting/macro/code organization issues that we are all been suffering, so it is a no-brainer (for me). But it is not about me. I want to avoid sole maintainer situation as much as you do. My intention is really to introduce MyDef to a point where fellow developers either appreciate the solution as much as I do or send me enough feedback to show me my blindsight that MyDef is not actually good. Even for this PR, my main intention is to demonstrate pros/cons and have something to discuss. If in the end, you are not convinced of the benefit, then I have no intention of merging it.

So far, my main feedback is unfamiliarity, which is fair feedback, so I am working on that ;) 

> 
>     4. Do downstream maintainers modify these scripts in their distributions? If so, will this change interfere with those modifications? I suppose they can workaround by just modifying the already generated files and adjusting their build scripts.

Exactly.
One of the good aspects of MyDef is it really emphasizes multi-layer aspects of development and emphasizes that the generated code should be as readable and as true as what a coder would write without MyDef. Think of MyDef as just a pre-autogen step or an editor tool that only help but does not intrude. The only issue is that a team should ensure a version of "true" source or bear
the burden of synchronization/backporting.

> 
> My overall take is that these scripts are rarely updated, and templating them in some way is a good idea. That said, MyDef is unfamiliar, meaning the templates will likely have a sole maintainer in practice. This is essentially the state of all the code generation scripts in MPICH, so it's something we are currently able to tolerate.

Unfamiliar is a fair feedback. The question is how easy to grok starting from unfamiliar. If you don't know Python, it may take a week to learn. If you don't know bash, it make take a day to learn. If you don't know emacs or vim, it may take a few hours to learn. Of course, they all take months or lifetime to get better. I really aimed for people to be able to start work with MyDef merely after working though simple examples with a page long tutorials.

The sole maintainer situation is bad and it is always my main intention to avoid it. But I think you have been over-estimating it in practice. For example, the Fortran binding and mpi_f08 are very much seem impenetrable but you have been maintaining it here and there. The python bindings are quite complex (well too bad, it is complicated) but I see you have been maintaining it here and there. And even external contributors have contributed patches. I understand a lot of these patches bring along the pain of learning unfamiliar code, but that has always been the life of maintainers. Initially, we are all being forced and reluctantly working on unfamiliar parts of code, but as we keep maintaining the code, they will become familiar to some degree. 
So I don't think the fear of new technology/solution is a legit factor, rather, the factors are 1. how easy to grok by unfamiliar developers. 2. how the solutions (new and old) compare once we are familiar with both.

hzhou committed:
mpicxx: configure to allow mpicxx skip libmpicxx

In case of --disable-cxx or --enable-mpi-abi or MPICXXLIBNAME is the
same as MPILIBNAME, we still need to install mpicxx to support cxx
compilers, but it should skip -lmpicxx.

hzhou committed:
env: whitespace changes to mpicc scripts

We'll verify the mydef script in the next commit does not introduce
unintended changes. Make some trivial whitespace adjustments in this
commit to facility verification of next commit using diff.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/ch3/most
test:mpich/custom
config: mpi-abi

✔️

hzhou committed:
configure: add --disable-bash option

Add an option for testing wrapper scripts using posix shell.

hzhou committed:
env: remove handling of quotes in sh version of scripts

Quoting inside a variable expansion ($allarg) won't work. The shell will
pass the quotes as literals and break the arguments at spaces.

Let's hope user will have a bash shell that support array or
they will never use arguments that contains spaces or quotes.

If we ever have a use case that can't work around, consider support
python or perl for wrapper scripts.

hzhou commented:
test:mpich/custom
config: nobash
